go/kbfs/simplefs/uploads.go

```diff
@@ -0,0 +1,250 @@
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package simplefs
+
+import (
+	"encoding/base64"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/kbfs/kbfscrypto"
+	"github.com/keybase/client/go/kbfs/libkbfs"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+)
+
+const (
+	// ulCtxOpID is the display name for the unique operation SimpleFS ID tag.
+	ulCtxOpID = "SFSULID"
+)
+
+// ulCtxTagKey is the type used for unique context tags
+type ulCtxTagKey int
+
+const (
+	// ulCtxIDKey is the type of the tag for unique operation IDs.
+	ulCtxIDKey ulCtxTagKey = iota
+)
+
+type upload struct {
+	dirToDelete *string
+	uploadID    string
+	opid        keybase1.OpID
+
+	state keybase1.UploadState
+}
+
+type uploadManager struct {
+	k         *SimpleFS
+	publisher libkbfs.SubscriptionManagerPublisher
+	cacheDir  string
+
+	lock      sync.RWMutex
+	tempPaths map[string]bool
+	uploads   map[string]upload
+}
+
+func newUploadManager(simpleFS *SimpleFS) *uploadManager {
+	return &uploadManager{
+		k:         simpleFS,
+		publisher: simpleFS.config.SubscriptionManagerPublisher(),
+		cacheDir:  simpleFS.config.KbEnv().GetCacheDir(),
+		tempPaths: make(map[string]bool),
+		uploads:   make(map[string]upload),
+	}
+}
+
+func (m *uploadManager) makeContext() (ctx context.Context) {
+	return libkbfs.CtxWithRandomIDReplayable(context.Background(), ulCtxIDKey, ulCtxOpID, m.k.log)
+}
+
+func (m *uploadManager) makeTempDir() (tempDirPath string, err error) {
+	buf := make([]byte, 12)
+	err = kbfscrypto.RandRead(buf)
+	if err != nil {
+		return "", err
+	}
+	cacheDirAbs, err := filepath.Abs(m.cacheDir)
+	if err != nil {
+		return "", err
+	}
+	tempDirPath = filepath.Join(cacheDirAbs, "uploads", base64.RawURLEncoding.EncodeToString(buf))
+	err = os.MkdirAll(tempDirPath, os.ModePerm)
+	if err != nil {
+		return "", err
+	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.tempPaths[tempDirPath] = true
+
+	return tempDirPath, nil
+}
+
+func (m *uploadManager) doWhileLocked(action func()) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	action()
+}
+
+func (m *uploadManager) getUpload(uploadID string) (upload, bool) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	u, ok := m.uploads[uploadID]
+	return u, ok
+}
+
+func (m *uploadManager) waitForDownload(uploadID string) {
+	upload, ok := m.getUpload(uploadID)
+	if !ok {
+		return
+	}
+	defer m.publisher.PublishChange(keybase1.SubscriptionTopic_UPLOAD_STATUS)
+
+	err := m.k.SimpleFSWait(m.makeContext(), upload.opid)
+	if err != nil {
+		m.doWhileLocked(func() {
+			upload, ok := m.uploads[uploadID]
+			if !ok {
+				return
+			}
+			if errors.Cause(err) == context.Canceled {
+				upload.state.Canceled = true
+			} else {
+				errStr := err.Error()
+				upload.state.Error = &errStr
+			}
+			m.uploads[uploadID] = upload
+		})
+		return
+	}
+	if upload.dirToDelete != nil {
```

Comment:
Seems like someone could misuse the API and use the same temporary directory for more than one upload, and this delete would break that second upload when the first one finishes.  Is it worth enforcing that each temp dir can only be used once?  Not sure how to do that most robustly though.  Also, in theory a temp dir could be a child of another temp dir, and when the latter upload finishes it could break the former upload.

Reply:
I'm not very worried about one temp dir being a child of another one. The temp dirs used by uploads are always created here, so that shouldn't happen right?

Re single temp dir, it's made clear in the avdl comment (after your suggestion). It seems good enough to me. WDYT?

Reply:
The directories are made on the Go side, but the API user can pass in whatever they want right?  So they could choose to use a subdir of a previous temp dir.  Obviously that would be a misuse of the API, but I was just brainstorming ways to harden the Go side of it.  It's fine with me if you don't think it's worth it.